### PR TITLE
Feature `SpectralCoord` implementation

### PR DESCRIPTION
**Pull request by nmearl**

### CODE REVIEW

File: astropy/coordinates/spectral_coordinate.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+    'optical': u.doppler_optical,
+    'relativistic': u.doppler_relativistic
+}
+
+__all__ = ['SpectralCoord']
+
+
+class SpectralCoord(u.Quantity):
+    """
+    Coordinate object representing a spectral axis.
```
Comment (astrofrog) [inline astropy/coordinates/spectral_coordinate.py]: Like SkyCoord, this doesn't have to represent just a spectral axis, just spectral values in general. So here and below I would remove/change references to 'axis'.

File: astropy/coordinates/spectral_coordinate.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+    Attributes
+    ----------
+    value : ndarray or `Quantity` or `SpectralCoord`
+        Spectral axis data values.
+    unit : str or `Unit`
+        Unit for the given data.
+    rest : `Quantity`, optional
+        The rest value to use for velocity space transformations.
+    observer : `BaseCoordinateFrame`, optional
+        The coordinate frame of the observer.
```
Comment (astrofrog) [inline astropy/coordinates/spectral_coordinate.py]: This is more than the frame, it's the actual position of the observer (i.e. a frame instance with data). So maybe rephrase to say 'The coordinates (position and velocity) of the observer'. In addition I wonder whether we should allow SkyCoord objects here? Same comments for the target below.

File: astropy/coordinates/spectral_coordinate.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    target : `BaseCoordinateFrame`, optional
+        The coordinate frame of the target.
+    """
+    def __new__(cls, value, unit=None, rest_value=None,
```
Comment (astrofrog) [inline astropy/coordinates/spectral_coordinate.py]: rest_value -> rest to be consistent with docstring - I think ``rest`` makes more sense since ``rest_value`` sounds like it should maybe not be a quantity.

File: astropy/coordinates/spectral_coordinate.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+                velocity_convention=None, observer=None, target=None,
+                **kwargs):
+        obj = super().__new__(cls, value, unit=unit, **kwargs)
+
+        obj.rest_value = rest_value or 0 * unit
```
Comment (astrofrog) [inline astropy/coordinates/spectral_coordinate.py]: I think it would make more sense to use ``None`` as the default so that we know whether it's been initialized or not, so we can return appropriate errors when trying to convert e.g. velocity to frequency? (applies to other instances of 0 * unit below)

File: astropy/coordinates/spectral_coordinate.py [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+    value : ndarray or `Quantity` or `SpectralCoord`
+        Spectral axis data values.
+    unit : str or `Unit`
+        Unit for the given data.
+    rest : `Quantity`, optional
+        The rest value to use for velocity space transformations.
+    observer : `BaseCoordinateFrame`, optional
+        The coordinate frame of the observer.
+    target : `BaseCoordinateFrame`, optional
+        The coordinate frame of the target.
```
Comment (astrofrog) [inline astropy/coordinates/spectral_coordinate.py]: ``velocity_convention`` is missing from the docstring

File: astropy/coordinates/spectral_coordinate.py [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+        obj.velocity_convention = velocity_convention or 'relativistic'
+        obj.observer = observer
+        obj.target = target
+
+        return obj
+
+    def __array_finalize__(self, obj):
+        super().__array_finalize__(obj)
+
+        if obj is not None:
```
Comment (astrofrog) [inline astropy/coordinates/spectral_coordinate.py]: Maybe add a comment to explain under what circumstances this is True or False.

File: astropy/coordinates/spectral_coordinate.py [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+                    0, 0, 0, unit=u.km/u.s)
+
+    @property
+    def observer(self):
+        """
+        The coordinate frame from which the observation was taken.
+
+        Returns
+        -------
+        : `BaseCoordinateFrame`
```
Comment (astrofrog) [inline astropy/coordinates/spectral_coordinate.py]: Don't include ``: `` at the start of the line, you can just give the type directly according to the numpydoc docs (applies to other instances below too)

File: astropy/coordinates/spectral_coordinate.py [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+        -------
+        : `Quantity`
+            Rest value as an astropy `Quantity` object.
+        """
+        return self._rest_value
+
+    @rest_value.setter
+    @u.quantity_input(value=['length', 'frequency', 'energy', 'speed'])
+    def rest_value(self, value):
+        self._rest_value = value
```
Comment (astrofrog) [inline astropy/coordinates/spectral_coordinate.py]: I think we might want to consider now allowing some of the setters, since we need to make it clear whether things are conversions or re-assignments. We do want to allow re-assignments though, so maybe this is the best way. Just putting this up for discussion :)

Comment (nmearl) [reply astropy/coordinates/spectral_coordinate.py]: > I think we might want to consider now allowing some of the setters

Did you mean "consider _not_ allowing some of the setters"? In this particular context, the `rest` property would act as either a setter (if we change the behavior of the initialized `SpectralCoord` as above to be `None` by default), and a converted (when the `rest` property has been defined). Do you think breaking this out as a separate method call (e.g. `with_rest(new_rest_quantity)` would be more descriptive?

Comment (astrofrog) [reply astropy/coordinates/spectral_coordinate.py]: Yes sorry I meant 'not' - a critical difference :) I think to convert to a different ``rest``, users should probably use ``to(...)`` with a different rest?

Comment (nmearl) [reply astropy/coordinates/spectral_coordinate.py]: That sounds reasonable. We still want to store the new `rest` value on the `SpectralCoord` object and allow `spec_coord.rest` to return it, yeah? Just require that it be set through `.to`? (This seems like it'll require some good documentation because it seems non-obvious.)
